### PR TITLE
--account-id flag

### PR DESCRIPTION
**Pull request by scotwk**
@kapilt - I only changed two resources (RDS and RDSSnapshot).  Just making sure, is this what you were going for?  If so, I'll proceed on with the rest.

I haven't explored all these yet, but from a simple grep, the other things that appear to use an account ID are:
 - account
 - elasticache
 - redshift
 - s3
 - iamaccess filter (unsure if this needs changing as it is a filter and not a resource)

### CODE REVIEW

**Comment by kapilt** [comment]
this entails a runtime failure if account id is not passed in, that's not good behavior. i'd like to give some grace on preserving extant semantics with a deprecation warning, in consideration of extant deploys.

**Comment by scotwk** [comment]
I switched to a deprecation warning, which is issued in the logger and also on STDERR.  There are a few spots that I didn't switch to the command line because I think they should continue using IAM.  

 - iamaccess filter
 - account resource

Are those correct?

File: c7n/query.py [hunk 1]
```diff
@@ -225,6 +229,28 @@ def augment(self, resources):
...
+        IAM.
+        """
+        if self.config.account_id:
+            return self.config.account_id
+
+        # Below this comment is deprecated.
+        msg = ("Warning: Inferring the account ID from IAM is deprecated. "
+               "Use the --account-id flag to specify this value.")
+        self.log.warning(msg)
+        print(msg, file=sys.stderr)
```
Comment (kapilt) [inline c7n/query.py]: we should probably use actual python stdlib deprecation framework.

Comment (scotwk) [inline c7n/query.py]: @kapilt - like so?

```
import warnings
warnings.warn("Some msg", DeprecationWarning)
```

Comment (kapilt) [inline c7n/query.py]: yup that will do in a nutshell, we want to capture policy name in the message. ideally we want to capture this in a validation as opposed to runtime output, but we need a bit more infrastructure on validation for that.

Comment (scotwk) [inline c7n/query.py]: @kapilt - Looks like `DeprecationWarning` is ignored by default.  I enabled it using the line below.  Do you want this in place?  (Seems like weird behavior that it has to be explicitly enabled)

```
warnings.simplefilter('always', DeprecationWarning)
```

As for validation - I did add the check at validation time for the S3 attach-encrypt action.  But you are right that it is a runtime problem for the others.

**Comment by scotwk** [comment]
@kapilt - This one should be good to go.  Removing the WIP designation.

Coverage decreased largely as a result of deleting covered lines of code.  There was some duplicate methods in some of the resources that I combined in their parent class.

**Comment by scotwk** [comment]
@kapilt - should be good to go now based on our phone conversation

**Comment by kapilt** [comment]
so i noticed there's an api to get this info unambiguously that works for all scenarios (iam user, sts role, instance role) via sts.GetCallerContext. ideally we can initialize this value if not passed in the cli, instead of hunting for the id in each of the resource managers. much like the default region lookup like now. it will need refactoring later for additional providers (provider.init), but that seems fine.

**Comment by scotwk** [comment]
@kapilt - just saw your comment at the end of last week and I'm getting back to it for the first time.  I don't fully understand what the request is.  I didn't find `sts.GetCallerContext`, although I did find `boto3.client('sts').get_caller_identity()`.  

Is the request to continue using the new `--acount-id` flag on the cli, but when it is not passed we will use sts instead of iam like we do now?  

If we are meeting tomorrow I can get clarification on that call.

**Comment by scotwk** [comment]
per phone call - get rid of account ID flag and infer it from sts

**Comment by kapilt** [comment]
effectively, we're talking about going to what this pr was originally doing, switch out to config.account_id, but with the value from sts by default.

**Comment by scotwk** [comment]
@kapilt - I implemented the new approach - did I get our approach correct?  Here is what I did:

 - remove the `--account-id` flag which had been added in a previous iteration of this PR
 - switch the `utils.get_account_from_iam` method to `utils.get_account_from_sts`

There are a ton of tests failing because we have to re-record them all using STS instead of IAM.  (That seems like it could be a big undertaking because we may have to setup the environment for those tests?)  I didn't want to spend time on it until I knew this was the approach you wanted.

**Comment by scotwk** [comment]
put the account-id back into the config on startup

**Comment by scotwk** [comment]
@kapilt - I moved the call to STS into `cli.py` and then it populates it into config.account_id which is used elsewhere.

**Comment by kapilt** [comment]
was about to merge and realized this is missing the lambda entry point as well

**Comment by scotwk** [comment]
Set account ID in handler.py (Config class) by doing an STS callout

**Comment by scotwk** [comment]
@kapilt - I haven't modified handler.py yet, so I think this will break lambdas?  (I am in the process of doing it right now.  I have the change, but it is breaking a test and I am debugging.)

**Pull request merged**